### PR TITLE
Changes to the reachability slicer

### PR DESCRIPTION
**Pull request by forejtv**
* Add forwards reachability to the slicer, in addition to the backwards reachability currently implemented.
* Add command-line options for the slicer in jbmc.

### CODE REVIEW

File: regression/cbmc-java/reachability-slice/test.desc
Comment (tautschnig) [inline regression/cbmc-java/reachability-slice/test.desc]: How about some assertion explicitly stated in the code rather than relying on coverage numbering? If you have to rely on coverage annotations, could you add a comment in the code to the block that you expect this to be mapped to?

File: src/goto-instrument/reachability_slicer.cpp
Comment (tautschnig) [inline src/goto-instrument/reachability_slicer.cpp]: Formatting changes only? If so, change it to a ranged-for.

File: src/goto-instrument/reachability_slicer.cpp
Comment (tautschnig) [inline src/goto-instrument/reachability_slicer.cpp]: If that's what's happening, why not use `grapht` functionality to accomplish it?

File: src/jbmc/jbmc_parse_options.cpp
Comment (tautschnig) [inline src/jbmc/jbmc_parse_options.cpp:777]: Please add the same functionality to the cbmc front-end.

File: regression/cbmc-java/reachability-slice/test2.desc
Comment (tautschnig) [inline regression/cbmc-java/reachability-slice/test2.desc]: Looking at the tests failing on AppVeyor, I am wondering whether `--show-goto-functions` actually does not produce any output on Windows.

File: regression/cbmc-java/reachability-slice/test.desc
Comment (forejtv) [reply regression/cbmc-java/reachability-slice/test.desc]: Assertion did not work here, surprisingly. The problem is that the assertion in Java throws an exception, so there's not really anything reachable after it, which makes the test vacuous. This is why I went for coverage criterion instead. I'll add a comment.

Comment (tautschnig) [reply regression/cbmc-java/reachability-slice/test.desc]: Ok, thanks!

File: src/goto-instrument/reachability_slicer.cpp
Comment (forejtv) [reply src/goto-instrument/reachability_slicer.cpp]: No idea. I was just modifying the code that was there already. I'd like to avoid making bigger changes, I don't know the codebase well enough to guarantee that I'm not doing more harm than good.

Comment (tautschnig) [reply src/goto-instrument/reachability_slicer.cpp]: I'd suggest the following:
1. Change `grapht::get_reachable` to take a `std::vector` or a `std::set` of `node_indext` as the first parameter.
2. Implement `fixedpoint_from_assertions` as follows:
````c++
for(const auto &cfg_entry : cfg.entry_map)
{
  std::vector<cfgt::node_indext> queue;
  if(criterion(cfg_entry.first) || is_threaded(cfg_entry.first))
    queue.push_back(cfg_entry.second);

  std::vector<cfgt::node_indext> reachable = cfg.get_reachable(queue, true);
  for(const auto index : reachable)
    cfg[index].reachable_from_assertion = true;
}
````

And then `reachability_slicert::fixedpoint_to_assertions` could also be rewritten in the same way, just setting the second argument of `get_reachable` to `false`.

File: src/goto-instrument/reachability_slicer.cpp
Comment (romainbrenguier) [inline src/goto-instrument/reachability_slicer.cpp]: use `const auto`

File: src/goto-instrument/reachability_slicer.cpp
Comment (romainbrenguier) [inline src/goto-instrument/reachability_slicer.cpp]: not sure `queue` is appropriate here since it's not used as a queue

File: src/util/graph.h
Comment (romainbrenguier) [inline src/util/graph.h:474]: I would suggest `std::vector<node_indext> src_vector({src})`

File: src/util/graph.h
Comment (romainbrenguier) [inline src/util/graph.h:492]: `src` is already passed as a copy, so there is no need for a second copy

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [reply src/goto-instrument/reachability_slicer.cpp]: Why `auto`? Type isn't repeated on rhs or particularly complicated. `const &` seems sensible though?

File: regression/cbmc-java/reachability-slice/test.desc
Comment (thk123) [inline regression/cbmc-java/reachability-slice/test.desc]: To check I've understood correctly, 1003-1005 are removed because we go up to the coverage goal on 1002 and everything after is redundant? So if we added --cover-must-terminate, we'd still expect 1004 to be sliced, but 1002, 3, 5 would be present?

Comment (thk123) [reply regression/cbmc-java/reachability-slice/test.desc]: That's what the second test is doing - got it :+1:

File: src/cbmc/cbmc_parse_options.cpp
Comment (thk123) [inline src/cbmc/cbmc_parse_options.cpp:806]: To simplify this being used in multiple places, you might pull out a `reachability_slice_buildert` that handles this logic in the constructor then can call `operator()` to execute the appropriate reachability slice

File: src/cbmc/cbmc_parse_options.cpp
Comment (thk123) [inline src/cbmc/cbmc_parse_options.cpp]: To allow re-use, please define these in `reachability_slicer.h`

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [inline src/goto-instrument/reachability_slicer.cpp]: Nit: Since body of for loop is multi-line please wrap in braces

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [inline src/goto-instrument/reachability_slicer.cpp]: Consider to make more explicit the transformation being done. 
```C++
std::copy_if(
  cfg.entry_map.begin(), 
  cfg.entry_map.end(), 
  std::back_inserter(queue),
  [&](entry) { return criterion(entry.first) || is_threaded(entry.first); });
```

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [inline src/goto-instrument/reachability_slicer.cpp]: Same comment as above - but consider just pulling out whatever you do into a function - though not sure what I'd call it.

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [inline src/goto-instrument/reachability_slicer.cpp]: TODO

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [inline src/goto-instrument/reachability_slicer.cpp]: Since code is same apart from `true`/`false` param on this call, consider pulling out the shared code to de-duplicate it

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [inline src/goto-instrument/reachability_slicer.cpp]: > with respect to criterion

What criterion?

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [inline src/goto-instrument/reachability_slicer.cpp]: I had to read the comment to know what this variable meant, consider `include_forward_instructions`?

**Review by thk123 — COMMENTED** [review]
I feel like this is a bit light on tests, I think there should be additionally:

 - a duplicate for cbmc to check integrated correctly
 - a run without a property set for both

**Comment by thk123** [comment]
This also would benefit with TG bump that integrates it as well to avoid the `parse_options` drifting apart. Non-blocking and happy to do this (particularly if you do the comments re: centralising the integration parts!)

File: src/util/graph.h
Comment (forejtv) [reply src/util/graph.h:492]: I find the current version better, it avoids potential problems if anybody uses the function in the future and does not expect the vector to change.

File: src/cbmc/cbmc_parse_options.cpp
Comment (forejtv) [reply src/cbmc/cbmc_parse_options.cpp:806]: This is well beyond my C++ knowledge. I'd be happy for somebody to take this PR over and do it, but I myself don't have enough knowledge to even understand what exactly you're suggesting :-)

File: src/goto-instrument/reachability_slicer.cpp
Comment (forejtv) [reply src/goto-instrument/reachability_slicer.cpp]: Yes, that's intended. I have no idea what the parameter does. If I don't put `\param is_threaded` in the comment, doxygen on Travis complains. So from my perspective, it's either this, or deleting the whole comment.

File: src/goto-instrument/reachability_slicer.cpp
Comment (romainbrenguier) [reply src/goto-instrument/reachability_slicer.cpp]: The type name is long, and we don't use vector specific functions on this object, it's just used as a container, we could for instance change the return type of get_reachable to a set. This is maybe a personal preference  so I'm not insisting. 
I think making it a reference does not change anything since `get_reachable` creates a new vector, maybe a c++ expert could confirm that?

File: src/util/graph.h
Comment (peterschrammel) [reply src/util/graph.h:492]: It's already passed as a copy, hence changes only affect the copy, which is invisible to the caller.

File: src/goto-instrument/reachability_slicer_class.h
Comment (peterschrammel) [inline src/goto-instrument/reachability_slicer_class.h:27]: Why isn't this const?

File: src/cbmc/cbmc_parse_options.h
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.h]: put these options into `OPT_REACHABILITY_SLICER` in `goto-instrument/reachability_slicer.h`

File: src/goto-instrument/reachability_slicer.cpp
Comment (peterschrammel) [inline src/goto-instrument/reachability_slicer.cpp]: `is_threadedt` overapproximates the set of instructions that might be executed concurrently. These cannot be safely sliced away because we lack information on the multithreaded control flow here (synchronisation).

File: src/goto-instrument/reachability_slicer.cpp
Comment (peterschrammel) [reply src/goto-instrument/reachability_slicer.cpp]: The compiler will optimise away the copy on return in this case. Using a `const &` would be C++03 style, but is unnecessary in C++11.

File: src/goto-instrument/reachability_slicer.cpp
Comment (forejtv) [reply src/goto-instrument/reachability_slicer.cpp]: As discussed in person, I'll leave this in the original form.

File: src/goto-instrument/reachability_slicer_class.h
Comment (forejtv) [reply src/goto-instrument/reachability_slicer_class.h:27]: I've tried that now, the compiler then complains about the call to `criterion(entry.first)`.

File: src/goto-instrument/reachability_slicer.cpp
Comment (romainbrenguier) [inline src/goto-instrument/reachability_slicer.cpp]: typo in concu*e*rrently

File: src/util/graph.h
Comment (forejtv) [reply src/util/graph.h:474]: Hmm, so appveyor did not like this, so I reverted to the original version.

**Comment by forejtv** [comment]
@thk123 @tautschnig @peterschrammel @romainbrenguier Thanks for your reviews! The PR should now be ready to be re-reviewed.

I'll squash the commits once everybody is happy with the code.

File: src/cbmc/cbmc_parse_options.cpp
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: I think the "currently buggy" is not very helpful to users - when exactly is it broken? If you insist on marking it in some way, I'd use "experimental".

**Review by tautschnig — APPROVED** [review]

File: src/cbmc/cbmc_parse_options.cpp
Comment (forejtv) [reply src/cbmc/cbmc_parse_options.cpp]: Fair point. Changed to "experimental". The reason I added this line at all was that the option is accepted already (before this PR), and I felt the need to sync the help message with the actual list of switches.

**Comment by tautschnig** [comment]
Further approval is required anyway, but just saying that before merging the commits should be squashed, please.

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [inline src/goto-instrument/reachability_slicer.cpp]: The comment for this and the above method are identical, I guess this should say "perform *forwards* depth first..."?

File: src/goto-instrument/reachability_slicer.cpp
Comment (thk123) [reply src/goto-instrument/reachability_slicer.cpp]: Did you spot this comment? Not critical but not applied.

**Review by thk123 — APPROVED** [review]
Comment needs updating but otherwise LGTM

**Review by peterschrammel — APPROVED** [review]

**Comment by tautschnig** [comment]
@forejtv @smowton @romainbrenguier There seems to be an unexpected interaction between the pull requests merged yesterday. Could you please review this?

**Comment by smowton** [comment]
@tautschnig looking at this now

**Comment by smowton** [comment]
@tautschnig fix: https://github.com/diffblue/cbmc/pull/1965

**Pull request merged**
